### PR TITLE
stylo: Traverse anonymous children when styling

### PR DESCRIPTION
**Pull request by bholley**
The corresponding Gecko changes are at https://bugzilla.mozilla.org/show_bug.cgi?id=1292662

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12911)

### CODE REVIEW

**Comment by bholley** [comment]
(Though let's wait to land this until the gecko-side changes are ready)

File: ports/geckolib/wrapper.rs [hunk 1]
```diff
@@ -349,6 +353,41 @@ impl<'ln> TNode for GeckoNode<'ln> {
...
+// like Servo does. However, for nodes with anonymous children, we use a custom
+// (heavier-weight) Gecko-implemented iterator.
+pub enum GeckoChildrenIterator<'a> {
+    Current(Option<GeckoNode<'a>>),
+    GeckoIterator(*mut bindings::StyleChildrenIterator),
+}
+
+impl<'a> Drop for GeckoChildrenIterator<'a> {
+    fn drop(&mut self) {
+        match *self {
```
Comment (emilio) [inline ports/geckolib/wrapper.rs]: nit: if let.

File: ports/geckolib/wrapper.rs [hunk 2]
```diff
@@ -349,6 +353,41 @@ impl<'ln> TNode for GeckoNode<'ln> {
...
+    type Item = GeckoNode<'a>;
+    fn next(&mut self) -> Option<GeckoNode<'a>> {
+        match *self {
+            GeckoChildrenIterator::Current(curr) => {
+                let next = curr.and_then(|node| node.next_sibling());
+                *self = GeckoChildrenIterator::Current(next);
+                curr
+            },
+            GeckoChildrenIterator::GeckoIterator(it) => unsafe {
+                Gecko_GetNextStyleChild(it).as_ref().map(|n| GeckoNode::from_ref(n))
```
Comment (emilio) [inline ports/geckolib/wrapper.rs]: You can eta-reduce this closure ;-)

**Comment by emilio** [comment]
r=me with those nits, awesome you also trimmed down the TNode trait :)

File: ports/geckolib/wrapper.rs (hunk 2 above)
Comment (bholley) [inline ports/geckolib/wrapper.rs]: I can't:

   Compiling geckoservo v0.0.1 (file:///files/mozilla/stylo/bb/servo/ports/geckolib)
wrapper.rs:376:54: 376:57 error: the trait bound `unsafe fn(&gecko_bindings::bindings::nsINode) -> wrapper::GeckoNode<'_> {wrapper::GeckoNode<'_>::from_ref}: std::ops::FnOnce<(&gecko_bindings::bindings::nsINode,)>` is not satisfied [E0277]
wrapper.rs:376                 Gecko_GetNextStyleChild(it).as_ref().map(GeckoNode::from_ref)
                                                                    ^~~
wrapper.rs:376:54: 376:57 help: run `rustc --explain E0277` to see a detailed explanation
error: aborting due to previous error
error: Could not compile `geckoservo`.

To learn more, run the command again with --verbose.

File: ports/geckolib/wrapper.rs (hunk 1 above)
Comment (bholley) [inline ports/geckolib/wrapper.rs]: Fixed.

**Comment by bholley** [comment]
@bors-servo p=1

Need this to rev the branch.

**Comment by bholley** [comment]
#8769

**Pull request merged**